### PR TITLE
test: introduce charm.py unittest coverage

### PR DESCRIPTION
**Pull request by marcoppenheimer**
#### `test: introduce charm.py and cluster.py unittest coverage`
- Due to active development, there were no unittests for `charm.py`. These have been added
- Added additional `cluster.py` and `tls.py` tests to add coverage for new features
## Review Notes
- Please focus on `tests/unit/test_charm.py`
- There are a lot of tests, so most help would be whether any were implemented incorrectly from an `ops.testing` standpoint

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:
             return
 
         # check whether restart is needed for all `*_changed` events
-        self.on[self.restart.name].acquire_lock.emit()
+        self.on[f"{self.restart.name}"].acquire_lock.emit()
```
Comment (zmraul) [inline src/charm.py:130]: Is this because of a change in RollingOps?

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+ops.testing.SIMULATE_CAN_CONNECT = True
+logger = logging.getLogger(__name__)
+
+CONFIG = str(yaml.safe_load(Path("./config.yaml").read_text()))
+ACTIONS = str(yaml.safe_load(Path("./actions.yaml").read_text()))
+METADATA = str(yaml.safe_load(Path("./metadata.yaml").read_text()))
+
+
+@pytest.fixture
+def harness():
```
Comment (zmraul) [inline tests/unit/test_charm.py:29]: Should you just add `peer_rel_id = harness.add_relation(PEER, CHARM_KEY)` here instead?

**Review by zmraul — APPROVED** [review]
That's a lot of testing.

File: tests/unit/test_charm.py (hunk 2 above)
Comment (marcoppenheimer) [reply tests/unit/test_charm.py:29]: I wanted to test the scenario where `if not self.cluster.relation' somwhere I think.

**Pull request merged**